tiledb/sm/filesystem/posix.cc

```diff
@@ -299,6 +300,60 @@ tuple<Status, optional<std::vector<directory_entry>>> Posix::ls_with_sizes(
   return {Status::Ok(), entries};
 }
 
+bool Posix::ls_cb(
+    const URI& uri, LsCallback cb, void* data, bool recursive) const {
+  std::string path = uri.to_path();
+  struct dirent** paths;
+  // scandir to traverse sorted objects, so we can invoke callback directly.
+  int path_num = scandir(path.c_str(), &paths, NULL, alphasort);
+  if (path_num < 0) {
+    throw std::runtime_error(
+        std::string("Cannot list files in directory '") + path + "'; " +
+        strerror(errno));
+  }
+  // Free memory allocated by scandir when we leave scope.
+  ScopedExecutor scoped_executor([&]() {
+    for (int i = 0; i < path_num; i++) {
+      free(paths[i]);
+    }
+    free(paths);
+  });
+
+  for (int i = 0; i < path_num; i++) {
+    auto next_path = paths[i];
+    if (!strcmp(next_path->d_name, ".") || !strcmp(next_path->d_name, "..")) {
+      continue;
+    }
+    std::string abspath = path + "/" + next_path->d_name;
+
+    // Getting the file size here incurs an additional system call
+    // via file_size() and ls() calls will feel this too.
+    // If this penalty becomes noticeable, we should just duplicate
+    // this implementation in ls() and don't get the size
+    if (next_path->d_type == DT_DIR) {
```

Comment:
```suggestion
    if (recursive && next_path->d_type == DT_DIR) {
```

I'm pretty sure we're missing a check on whether this is a recursive `ls` here.

Reply:
Which also means we probably don't have tests for the non-recursive `ls_cb` which we should probably add.